src/hashset.h

```diff
@@ -0,0 +1,205 @@
+    unsigned long maxChainLen;   /* probing chain length */
+    unsigned long totalChainLen; /* buckets with probing flag */
+    unsigned long htSize;        /* buckets * positions-per-bucket */
+    unsigned long htUsed;        /* num elements */
+    unsigned long *clvector;
+} hashsetStats;
+
+/* --- Prototypes --- */
+
+/* Hash function (global seed) */
+void hashsetSetHashFunctionSeed(const uint8_t *seed);
+uint8_t *hashsetGetHashFunctionSeed(void);
+uint64_t hashsetGenHashFunction(const char *buf, size_t len);
+uint64_t hashsetGenCaseHashFunction(const char *buf, size_t len);
+
+/* Global resize policy */
+void hashsetSetResizePolicy(hashsetResizePolicy policy);
+
+/* Hashset instance */
+hashset *hashsetCreate(hashsetType *type);
+void hashsetRelease(hashset *t);
+void hashsetEmpty(hashset *t, void(callback)(hashset *));
+hashsetType *hashsetGetType(hashset *t);
+void *hashsetMetadata(hashset *t);
+size_t hashsetSize(hashset *t);
+size_t hashsetBuckets(hashset *t);
+size_t hashsetProbeCounter(hashset *t, int table);
+size_t hashsetMemUsage(hashset *t);
+void hashsetPauseAutoShrink(hashset *t);
+void hashsetResumeAutoShrink(hashset *t);
+int hashsetIsRehashing(hashset *t);
+int hashsetIsRehashingPaused(hashset *t);
+void hashsetRehashingInfo(hashset *t, size_t *from_size, size_t *to_size);
+int hashsetRehashMicroseconds(hashset *s, uint64_t us);
+int hashsetExpand(hashset *t, size_t size);
+int hashsetTryExpand(hashset *t, size_t size);
+int hashsetExpandIfNeeded(hashset *t);
+int hashsetShrinkIfNeeded(hashset *t);
+hashset *hashsetDefragInternals(hashset *t, void *(*defragfn)(void *));
+
+/* Elements */
+int hashsetFind(hashset *t, const void *key, void **found);
+void **hashsetFindRef(hashset *t, const void *key);
+/* void *hashsetFetchElement(hashset *t, const void *key); */
+int hashsetAdd(hashset *t, void *elem);
+int hashsetAddOrFind(hashset *t, void *elem, void **existing);
+void *hashsetFindPositionForInsert(hashset *t, void *key, void **existing);
+void hashsetInsertAtPosition(hashset *t, void *elem, void *position);
+int hashsetReplace(hashset *t, void *elem);
+int hashsetPop(hashset *t, const void *key, void **popped);
+int hashsetDelete(hashset *t, const void *key);
+void **hashsetTwoPhasePopFindRef(hashset *t, const void *key, void **position);
+void hashsetTwoPhasePopDelete(hashset *t, void *position);
+
+/* Iteration & scan */
+size_t hashsetScan(hashset *t, size_t cursor, hashsetScanFunction fn, void *privdata, int flags);
+void hashsetInitIterator(hashsetIterator *iter, hashset *t);
+void hashsetInitSafeIterator(hashsetIterator *iter, hashset *t);
+void hashsetResetIterator(hashsetIterator *iter);
+hashsetIterator *hashsetCreateIterator(hashset *t);
+hashsetIterator *hashsetCreateSafeIterator(hashset *t);
+void hashsetReleaseIterator(hashsetIterator *iter);
+int hashsetNext(hashsetIterator *iter, void **elemptr);
+#endif
+
+/* Random elements */
+int hashsetRandomElement(hashset *t, void **found);
```

Comment:
```suggestion
int hashsetRandomElement(hashset *hs, void **found);
```
I take some objection to having the default "one character" abbreviation for a hashset to be t. I would rather use s for set. or hs for hashset. I feel like there are too many other potential conflicts like target or type.

Reply:
You're right, it was `t` for "table" for a while, before we settle on the name "hashset". I was going to change all to `s`, but if you prefer `hs`, we can use that instead. Which one do you prefer of s and hs?

Reply:
I picked `s`.